pkg/xds/sync/egress_proxy_builder.go

```diff
@@ -3,60 +3,144 @@ package sync
 	ReadOnlyResManager manager.ReadOnlyResourceManager
 	LookupIP           lookup.LookupIPFunc
 	MetadataTracker    DataplaneMetadataTracker
+	meshCache          *xds_cache.Cache
+	DataSourceLoader   datasource.Loader
 
-	apiVersion envoy.APIVersion
-}
-
-func filterMeshesByMTLS(rs core_model.Resource) bool {
-	return rs.(*core_mesh.MeshResource).Spec.GetMtls().GetEnabledBackend() != ""
+	zone       string
+	apiVersion envoy_common.APIVersion
 }
 
-func (p *EgressProxyBuilder) build(key core_model.ResourceKey) (*xds.Proxy, error) {
-	ctx := context.Background()
+func (p *EgressProxyBuilder) Build(key core_model.ResourceKey) (*xds.Proxy, error) {
+	ze := core_mesh.NewZoneEgressResource()
+	if err := p.ReadOnlyResManager.Get(p.ctx, ze, core_store.GetBy(key)); err != nil {
+		return nil, err
+	}
 
-	zoneEgress := core_mesh.NewZoneEgressResource()
-	if err := p.ReadOnlyResManager.Get(ctx, zoneEgress, core_store.GetBy(key)); err != nil {
+	var meshList core_mesh.MeshResourceList
+	if err := p.ReadOnlyResManager.List(
+		p.ctx,
+		&meshList,
+	); err != nil {
 		return nil, err
 	}
 
-	var meshes core_mesh.MeshResourceList
-	if err := p.ReadOnlyResManager.List(ctx, &meshes, core_store.ListByFilterFunc(filterMeshesByMTLS)); err != nil {
+	var meshes []*core_mesh.MeshResource
+	for _, mesh := range meshList.Items {
+		if mesh.MTLSEnabled() {
+			meshes = append(meshes, mesh)
+		}
+	}
+
+	var zoneIngressesList core_mesh.ZoneIngressResourceList
+	if err := p.ReadOnlyResManager.List(
+		p.ctx,
+		&zoneIngressesList,
+	); err != nil {
 		return nil, err
 	}
 
-	var externalServices []*core_mesh.ExternalServiceResource
-	for _, mesh := range meshes.Items {
+	var zoneIngresses []*core_mesh.ZoneIngressResource
+	for _, zoneIngress := range zoneIngressesList.Items {
+		if zoneIngress.IsRemoteIngress(p.zone) {
+			zoneIngresses = append(zoneIngresses, zoneIngress)
+		}
+	}
+
+	externalServiceMap := map[string][]*core_mesh.ExternalServiceResource{}
+	meshEndpointMap := map[string]xds.EndpointMap{}
+	trafficRouteMap := map[string][]*core_mesh.TrafficRouteResource{}
+
+	for _, mesh := range meshes {
 		meshName := mesh.GetMeta().GetName()
 
-		var es core_mesh.ExternalServiceResourceList
-		if err := p.ReadOnlyResManager.List(ctx, &es, core_store.ListByMesh(meshName)); err != nil {
+		meshExternalServiceMap := map[string]*core_mesh.ExternalServiceResource{}
+
+		meshCtx, err := p.meshCache.GetMeshContext(p.ctx, syncLog, meshName)
+		if err != nil {
 			return nil, err
 		}
 
-		externalServices = append(externalServices, es.Items...)
+		meshResources := meshCtx.Resources
+		trafficPermissions := meshResources.TrafficPermissions()
+		trafficRoutes := meshResources.TrafficRoutes().Items
+		externalServices := meshResources.ExternalServices()
+
+		trafficRouteMap[meshName] = trafficRoutes
+
+		for _, dp := range meshCtx.DataplanesByName {
+			if !dp.Spec.Matches(map[string]string{
+				mesh_proto.ZoneTag: p.zone,
+			}) {
+				continue
+			}
+
```

Comment:
I think this whole logic of trying to match traffic permission to external service should be gone. With ZoneEgress we want to implement this with inbound Envoy RBAC filter.

We had this logic before because we had to filter out allowed ExternalServices from the client data plane proxy, but now we can do this in a better, more secure way.

Reply:
addressed